libcontainer/cgroups/fs2/create.go

```diff
@@ -95,10 +102,18 @@ func CreateCgroupPath(path string, c *configs.Cgroup) (Err error) {
 					}
 				}()
 			}
-			// Write cgroup.type explicitly.
-			// Otherwise ENOTSUP may happen.
-			cgType := filepath.Join(current, "cgroup.type")
-			_ = ioutil.WriteFile(cgType, []byte("threaded"), 0644)
+			// There are 4 types: 'domain', 'threaded', 'domain threaded', and 'domain invalid'.
+			// If got 'domain invalid', we should check whether the current config contains domain controller or not.
+			// If it does not contain domain controller, we can write "threaded" without retuning error.
+			cgTypeFile := filepath.Join(current, "cgroup.type")
+			cgType, _ := ioutil.ReadFile(cgTypeFile)
+			if strings.TrimSpace(string(cgType)) == "domain invalid" {
```

Comment:
One other thing we should consider is to give an error early if we are trying to join a cgroup in `threaded` mode but have domain controllers enabled in the configuration. That way we can avoid weird error messages and help users understand why cgroup setup is failing.

Reply:
Yes, I'll find the best location to check and give the error to user.

Reply:
You can do it right here, maybe something more like this (I've also cleaned up the error messages and comments):

```
cgTypeFile := filepath.Join(current, "cgroup.type")
cgType, _ := ioutil.ReadFile(cgTypeFile)
switch strings.TrimSpace(string(cgType)) {
// If the cgroup is in an invalid mode (usually this means there's an internal
// process in the cgroup tree, because we created a cgroup under an
// already-populated-by-other-processes cgroup), then we have to error out if
// the user requested controllers which are not thread-aware. However, if all
// the controllers requested are thread-aware we can simply put the cgroup into
// threaded mode.
case "domain invalid":
	if containsDomainController(c) {
		return fmt.Errorf("cannot enter cgroupv2 %q with domain controllers -- it is in an invalid state", current)
	} else {
		// Not entirely correct (in theory we'd always want to be a domain --
		// since that means we're a properly delegated cgroup subtree) but in
		// this case there's not much we can do and it's better than giving an
		// error.
		_ = ioutil.WriteFile(cgTypeFile, []byte("threaded"), 0644)
	}
// If the cgroup is in threaded mode, we can only use thread-aware controllers
// (and you cannot usually take a cgroup out of threaded mode).
case "threaded":
	if containsDomainController(c) {
		return fmt.Errorf("cannot enter cgroupv2 %q with domain controllers -- it is in threaded mode", current)
	}
}
```

Reply:
add `case "domain threaded":`

Reply:
I *think* we can safely use domain controllers in a `domain threaded` cgroup because it's the cgroup *above* a threaded cgroup. But honestly I'm not too sure -- I need to set up an Ubuntu 20.04 VM to test this stuff...

Reply:
Domain controllers can’t be used in a domain threaded cgroup. I have tested it in Fedora 31.

Reply:
Ah okay, thanks for testing it.